### ISSUE TITLE
`helm-resume` does not restore keymap

### ISSUE THREAD

**Issue by xiongtx** [emacs-helm/helm#1670]
## Expected behavior

`helm-resume` should restore the keymap associated with the previous `helm` call.

## Actual behavior from emacs-helm.sh if possible

`helm-resume` does not restore the keymap.

If `:keymap` is specified for the source instead of the `helm` call, it *is* restored. This is fine for my use case, but it's still surprising that the one keymap is restored but another is not.

## Steps to reproduce (recipe)

1. Evaluate the following:

```emacs-lisp
(helm
 :sources '((name . "test")
            (candidates . (a b c)))
 :buffer "*Helm test*"
 :keymap (let ((map (copy-keymap helm-map)))
           (define-key map (kbd "C-c n") (lambda ()
                                           (interactive)
                                           (message "Keymap is working!")))
           map))
```

2. Hit `C-c n`
3. See that the echo area displays `Keymap is working!`
4. Quit Helm
5. Call `helm-resume`
6. Hit `C-c n` again
7. See that `C-c n` is undefined

## Backtraces if some (M-x toggle-debug-on-error)

N/A

## Describe versions of helm, emacs, operating system etc.

- Helm: 2.4.0
- Emacs: 25.1.1
- OS: Fedora 24

**Comment by thierryvolpiatto** [comment]
Tianxiang Xiong <notifications@github.com> writes:

> Expected behavior
>
> helm-resume should restore the keymap associated with the previous helm-call.
>
> Actual behavior from emacs-helm.sh if possible
>
> helm-resume does not restore the keymap.
>
> Steps to reproduce (recipe)
>
>  1. Evaluate the following:
>
> (helm
>  :sources '((name . "test")
>             (candidates . (a b c)))
>  :buffer "*Helm test*"
>  :keymap (let ((map (copy-keymap helm-map)))
>            (define-key map (kbd "C-c n") (lambda ()
>                                            (interactive)
>                                            (message "Keymap is working!")))
>            map))

You should store your keymap inside the source to be resumed properly:

    (helm
     :sources (helm-build-sync-source "test"
                :candidates '(a b c)
                :keymap (let ((map (copy-keymap helm-map)))
                          (define-key map (kbd "C-c n") (lambda ()
                                                          (interactive)
                                                          (message "Keymap is working!")))
                          map))
     :buffer "*Helm test*")

Keymaps attributes are handled specially in helm due to the different
keymaps in each sources, the other minor-modes interferring etc...

The :keymap slot used in `helm` (the function) call have been introduced
long time ago when helm was able to handle one and only one keymap
(helm-map).
As soon as helm started, helm-map was made buffer-local with this value,
which was ok when we had only one keymap to handle, this would be ok
actually if each keymap was completely differents, but each keymap is
inheriting from helm-map, and this helm-map should not be modified at
startup, otherwise all keymap will inherit of the modified keys unless
they are source local.

So it is kept here for backward compatibility but should not be used when
you want your keymap restored in helm-resume.
I will add a note in helm documentation about this.

Thanks.

PS: When sending examples use helm-build-*source or helm-make-source, to
not confuse new users reading this about how to build sources.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
While I was at it I found two other issues with this example:

```elisp
(helm
 :sources `(,(helm-build-sync-source "test1"
                                     :candidates '(d e f))
             ,(helm-build-sync-source "test2"
                                      :candidates '(a b c)
                                      :keymap (let ((map (copy-keymap helm-map)))
                                                (define-key map (kbd "C-c n") (lambda ()
                                                                                (interactive)
                                                                                (message "Keymap is working!")))
                                                map)))
 :buffer "*Helm test*")
```
- When switching from the source with a keymap (test2) to test1, I expect the keymap in use is helm-map and C-c n is undefined, it was not the case, the keymap of test2 was reused.

- When entering "a" at prompt I expected C-c n was available, it was not the case until I hit any key (or C-c n twice).

This two issue have been fixed in commit d8708168.

**Comment by xiongtx** [comment]
Good, at least this issue managed to uncover some other problems!
